src/instructions/nor.rs

```diff
@@ -0,0 +1,410 @@
+use crate::circuit_io_type::CircuitIOType;
+use anyhow::{bail, Ok, Result};
+pub use ark_r1cs_std::boolean::AllocatedBool;
+use ark_r1cs_std::{
+    prelude::{AllocVar, Boolean},
+    R1CSVar,
+};
+use indexmap::IndexMap;
+use simpleworks::{gadgets::traits::BitwiseOperationGadget, marlin::ConstraintSystemRef};
+pub use CircuitIOType::{SimpleBoolean, SimpleUInt16, SimpleUInt32, SimpleUInt64, SimpleUInt8};
+
+pub fn nor(
+    operands: &IndexMap<String, CircuitIOType>,
+    constraint_system: ConstraintSystemRef,
+) -> Result<CircuitIOType> {
+    match operands
+        .values()
+        .collect::<Vec<&CircuitIOType>>()
+        .as_slice()
+    {
+        [SimpleBoolean(left_operand), SimpleBoolean(right_operand)] => {
+            let a = AllocatedBool::new_witness(constraint_system.clone(), || left_operand.value())?;
+            let b = AllocatedBool::new_witness(constraint_system, || right_operand.value())?;
+            let result = AllocatedBool::nor(&a, &b)?;
+            let result = Boolean::from(result);
+            Ok(SimpleBoolean(result))
```

Comment:
Is there any specific reason why we need to allocate new witnesses here? 
Or could we do something like this: `let result = left_operand.or(right_operand)?.not();`

I don't mind either way but there might be something I'm missing

Reply:
I don't think that it is necessary, if I'm not mistaken the allocation could be skipped and `a` and `b` could be the operand values.

Reply:
I imagine this has to do with the fact that the operands are `Boolean`s but `nor` is only implemented for `AllocatedBool`s. I think the correct thing to do here is to `match` the three possible `Boolean` values: `Is(AllocatedBool`, `Not(AllocatedBool)` and `Constant`. If they are allocated bools you use the internal value to perform `nor`, if they are constants you just do the `nor` without even generating constraints.

Reply:
> I imagine this has to do with the fact that the operands are `Boolean`s but `nor` is only implemented for `AllocatedBool`s. I think the correct thing to do here is to `match` the three possible `Boolean` values: `Is(AllocatedBool`, `Not(AllocatedBool)` and `Constant`. If they are allocated bools you use the internal value to perform `nor`, if they are constants you just do the `nor` without even generating constraints.

Yes, that was exactly the reason why I used the `AllocatedBools`, I saw the `nor` implementation and I tried to use it but we can use the `or` and `not` to do the same with `Booleans` and it should be all right. I did that change so let me know if it's okay.

Reply:
Yup, looks good to me 👍